src/libproc_macro/lib.rs

```diff
@@ -212,12 +213,62 @@ impl Span {
         ::__internal::with_sess(|(_, mark)| Span(mark.expn_info().unwrap().call_site))
     }
 
+    /// The original source file into which this span points.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn source_file(&self) -> Option<PathBuf> {
+        // If this span's token was produced by a macro
+        if self.0.ctxt() != SyntaxContext::empty() {
+            return None
+        }
+
+        let loc = __internal::lookup_char_pos(self.0.lo());
+
+        PathBuf::from(&loc.file.name).canonicalize().ok()
```

Comment:
This could actually return `&Path` if we don't care about canonicalization.

Reply:
Hmm, I would lean against canonicalizing, but I don't have that strong of an opinion.

Reply:
Well, it'll save an allocation to skip it so that's good enough for me.